insight-rest-client/src/main/java/org/nuxeo/ai/sdk/rest/api/DedupCaller.java

```diff
@@ -0,0 +1,138 @@
+import org.nuxeo.ai.sdk.rest.client.API;
+import org.nuxeo.ai.sdk.rest.client.InsightClient;
+import org.nuxeo.ai.sdk.rest.exception.InvalidEndpointException;
+import org.nuxeo.ai.sdk.rest.exception.InvalidParametersException;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
+
+public class DedupCaller implements Resource {
+
+    private static final TypeReference<List<String>> LIST_TYPE_REFERENCE = new TypeReference<List<String>>() {
+    };
+
+    private final Logger log = LogManager.getLogger(DedupCaller.class);
+
+    private final InsightClient client;
+
+    private final API.Dedup type;
+
+    public DedupCaller(InsightClient client, API.Dedup type) {
+        this.client = client;
+        this.type = type;
+    }
+
+    @Nullable
+    @Override
+    public <T> T call(Map<String, Serializable> parameters) throws IOException {
+        return call(parameters, null);
+    }
+
+    @Nullable
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T> T call(Map<String, Serializable> parameters, Serializable payload) throws IOException {
+        switch (this.type) {
+        case INDEX:
+            return (T) handleIndex(parameters, (TensorInstances) payload);
+        case FIND:
+            return (T) handleFind(parameters, (TensorInstances) payload);
+        default:
+            throw new InvalidEndpointException("No such endpoint " + this.type.name());
+        }
+    }
+
+    private Boolean handleIndex(Map<String, Serializable> parameters, TensorInstances payload)
+            throws JsonProcessingException {
+        String docId = (String) parameters.get(UID);
+        String xpath = (String) parameters.getOrDefault(XPATH_PARAM, DEFAULT_XPATH);
+        int distance = (int) parameters.getOrDefault(DISTANCE_PARAM, 0);
+        if (StringUtils.isAnyEmpty(docId, xpath) || payload == null) {
+            throw new InvalidParametersException("Document UUID, XPath and Payload are required parameters");
+        }
+
+        String json = MAPPER.writeValueAsString(payload);
+        return client.post(this.type.toPath(client.getProjectId(), docId, xpath, distance), json, response -> {
+            if (response.isSuccessful()) {
+                log.debug("Successfully indexed document {} with xpath {}", docId, xpath);
+                return true;
+            }
+
+            log.error(
+                    "Failed to index document {} with xpath {} for project {};\nURL: {}\nresponse code: {}\nmessage: {}",
+                    docId, xpath, client.getProjectId(), client.getUrl(), response.code(), response.message());
+            return false;
+        });
+    }
+
+    private List<String> handleFind(Map<String, Serializable> parameters, TensorInstances payload)
+            throws JsonProcessingException {
+        String docId = (String) parameters.get(UID);
+        String xpath = (String) parameters.getOrDefault(XPATH_PARAM, DEFAULT_XPATH);
```

Comment:
I really would prefer to always force a XPATH at this level. Define a default on ai-core if needed, but this stay open.

Reply:
ok